crates/ruff_linter/src/rules/ruff/rules/legacy_form_pytest_raises.rs

```diff
@@ -0,0 +1,294 @@
+                indented,
+                stmt.range(),
+            )));
+        }
+    }
+    checker.report_diagnostic(diagnostic);
+}
+
+fn try_fix_legacy_call(
+    context_type: PytestContextType,
+    stmt: &Stmt,
+    semantic: &SemanticModel,
+) -> Option<StmtWith> {
+    match stmt {
+        Stmt::Expr(StmtExpr { value, .. }) => {
+            let call = value.as_call_expr()?;
+
+            if PytestContextType::from_expr_name(&call.func, semantic) == Some(context_type) {
+                generate_with_statement(context_type, call, None, None, None)
+            } else if let PytestContextType::Raises = context_type {
+                // Handle call to match right after the legacy call
+                // This applies only for `raises` because it returns `ExceptionInfo`
+                // while `warns` and `deprecated_call` return what `func` returns
+                let inner_raises_call = call
+                    .func
+                    .as_attribute_expr()
+                    .filter(|expr_attribute| &expr_attribute.attr == "match")
+                    .and_then(|expr_attribute| expr_attribute.value.as_call_expr())
+                    .filter(|inner_call| {
+                        PytestContextType::from_expr_name(&inner_call.func, semantic)
+                            == Some(PytestContextType::Raises)
+                    })?;
+                let match_arg = call.arguments.args.first();
+                generate_with_statement(context_type, inner_raises_call, match_arg, None, None)
+            } else {
+                None
+            }
+        }
+        Stmt::Assign(ast::StmtAssign {
+            range: _,
+            targets,
+            value,
+        }) => {
+            let call = value.as_call_expr().filter(|call| {
+                PytestContextType::from_expr_name(&call.func, semantic) == Some(context_type)
+            })?;
+            let (optional_vars, assign_targets) = match context_type {
+                PytestContextType::Raises => {
+                    let [target] = targets.as_slice() else {
+                        return None;
+                    };
+                    (Some(target), None)
+                }
+                PytestContextType::Warns | PytestContextType::DeprecatedCall => {
+                    (None, Some(targets.as_slice()))
+                }
+            };
+
+            generate_with_statement(context_type, call, None, optional_vars, assign_targets)
+        }
+        _ => None,
+    }
+}
+
+fn generate_with_statement(
+    context_type: PytestContextType,
+    legacy_call: &ast::ExprCall,
+    match_arg: Option<&Expr>,
+    optional_vars: Option<&Expr>,
+    assign_targets: Option<&[Expr]>,
+) -> Option<StmtWith> {
+    let expected = if let Some((name, position)) = context_type.expected_arg() {
+        Some(legacy_call.arguments.find_argument_value(name, position)?)
+    } else {
+        None
+    };
```

Comment:
I think we could use `and_then` here (but haven't tested it):

```suggestion
    let expected = context_type.expected_arg().and_then(|(name, position)| legacy_call.arguments.find_argument_value(name, position));
```

Reply:
I think we can't use `and_then` here because we need different behavior:
For `raises`/`warns`, the `?` operator returns early if the argument isn't found.
For `deprecated_call`, we expect `None` and continue.